A stab at refining imported-from

lierdakil opened the pull request:
Continued from https://github.com/DanielG/ghc-mod/pull/823, https://github.com/DanielG/ghc-mod/commit/3f282a8cf9bab89974234de75cf10b83fc5c7b63

lierdakil committed:
Rough sketch for imported-from idea

lierdakil committed:
Minor fix to rough sketch for imported-from idea

lierdakil committed:
Refine imported-from

lierdakil committed:
Imported-from spec update

lierdakil commented:
@carlohamalainen, I refined the code a bit (a lot actually). It should work as expected on your examples (at least more or less).

It now _tries_ to check for haddock documentation, but if there are no interface files, it just bails and falls back to just exports visible to GHC. Less precise, but it works without haddock.

Also made symbol optional. It's only used for selecting preferred import -- I'm pretty confident in selecting smallest source span; although, again, `findSpanName` probably could be written a bit better -- I'm a bit rusty on generics, and a bit too lazy at the moment to wrap my head around them all over again.

Error reporting probably could also be better: for now on most reasonably probable scenarios it just outputs 'Nothing found'.

lierdakil commented:
P.S. For now, only works with 7.10 -- there's little changes required with 8.0 and a bit more with 7.8. I know about this, I will fix when able.

Also note there is a small problem with outputted package name: I think first field should print origin package/module? If so, then package isn't necessarily origin package. F.ex., primitive types like `Int` are defined in `ghc-prim`, but current code will show `base` in most cases, since it's the package those types are (usually) imported from. There's a simple fix for that, but I'm out of time for the moment.

DanielG commented:
I have no problem with dropping 7.8 so we can do that if porting is too much effort.

lierdakil commented:
Nah, it's fine, just a couple functions have different signatures.

lierdakil committed:
[imported-from] Show originating package name

lierdakil committed:
[imported-from] ghc8 compat

lierdakil committed:
[imported-from] ghc-7.8 compat

lierdakil committed:
[imported-from] compat cleanup

lierdakil committed:
[imported-from] comment on `gfindtype` in Gap (ghc 7.8 compat)

lierdakil committed:
[imported-from] Clean `findSpanName`

lierdakil committed:
[imported-from] Allow outputting several names

Case when this is relevant: consider a query for source point like this:
something   other
          ^--- queryed here
Which one do we choose, former or latter? I would argue we choose both
and let the client decide.

lierdakil committed:
[imported-from] Remove redundant import

lierdakil committed:
[imported-from] Fix some things with `showOutput`

Apparently, ghc 7.8 can be picky when to give you a package version.
It can return just empty version for wired-in packages like base, etc.
So I suppose we have to guess those from haddock file path if we can...

lierdakil committed:
[imported-from] Fix broken test

lierdakil commented:
I'm thinking at this point it could be a good idea to merge this back into `imported-from` branch, and continue from there @carlohamalainen, any objections?

carlohamalainen commented:
@lierdakil yep, go for it.
